### PR TITLE
feat: create SymbolIterator for block parsing

### PR DESCRIPTION
**Pull request by mhatzl**
This PR adds a `SymbolIterator` to iterate over the `Symbol` slice returned from the `Scanner`.
It enables convenient tokenization and parsing for block elements,
by allowing to nest iterators, and define functions to match for block end and to strip line prefixes.

This is part of preparation for parsing more complex elements like lists.

# Decisions

1. Nesting iterators

   Because block elements may allow nesting, iterators may also be nested
   to make parsing of elements nesting agnostic.
   Parent iterators only forward symbols to children, if the parent is not at the end,
   and an optional prefix is stripped.

   Iterators keep track of their nesting depth, to allow nesting of sequences like `(outer (inner))`.

2. Functions/Closures for end/prefix matching

   It was decided to allow functions to be passed for end/prefix matching instead of multiple Symbol 
   slices. This allows fine grained matching, and allows to distinguish between peeked and consumed 
   matching.

   Prefix matching is always consuming, because nested iterators must not see prefix symbols of parent 
   elements.

# Future work

- Use `SymbolIterator` for inlines

  Currently, inline parsing is done on a symbol slice.
  Inline parsing internally uses `get_symbol(index)` quite often, which makes it hard to adopt the 
  iterator approach. For now, blocks retrieve a `Symbol` vector for inline content,
  and then pass this to the inline parser.

  With the new iterator approach supporting end detection, it might be possible to use
  the same tokenization and parsing approach as done for blocks.
  The handling of ambiguous tokens like `***` won't be trivial, but I am confident that
  it can be done by also matching for Bold and Italic directly in the ambiguous parser.

  Changing inline parsing will take time, and possibly break things.
  I would suggest that we first implement more blocks to see if the iterator approach is feasible.
  If the new approach seems to work well for blocks, we can try to adopt it also for inlines.

- Iterators all the way

  The scanner creates the vector of `Symbol`s using the grapheme segmenter of `icu`.
  This segmenter itself is an iterator, so we might be able to directly embed this iterator
  in the `SymbolIteratorRoot`. This would require some form of a symbol cache to allow peeking,
  but using an internal `VecDeque` should do the trick.

  **Note:** This effectively removes the option to get a slice of remaining symbols,
  and therefore the option to get size hints of remaining symbols.
  The size hint could still be approximated by using the utf8 length of the input string.
  This won't be accurate on Windows, because new lines have utf8 length of two due to `\r\n`.

  I am also unsure about the effect on performance and memory usage, when switching to full iterator.
  Memory should in theory become better, because we do not allocate memory for all symbols.
  Compared to allocation needed for the transformed content, this improvement might be neglectable 
  though.

### CODE REVIEW

**Comment by mhatzl** [comment]
I had to fix `icu` dependencies and check in `Cargo.lock`, because `icu` has internal version conflicts between `1.2.0` and `1.3.0`.

Clippy currently fails due to old code.
This is fixed in PR #105 so I will merge it once this PR is merged into `main`.

**Comment by mhatzl** [comment]
I also implemented rendering for newline and whitespace, because I wanted to make sure the resulting HTML is still correct using the iterator approach.

**Comment by mhatzl** [comment]
After seeing the failed actions due to `icu` dependencies yesterday,
I updated to icu `1.3.0`, which allows us to use the grapheme iterator without any generated data.
This also allows to remove the pin to a specific version, because we only need `icu_segmenter` and `icu_locid` both of which probably won't create version conflicts.

**Review by nfejzic — CHANGES_REQUESTED** [review]
This is a big one, and big ones always have lots of comments. Some might be just do discuss a little bit and require no action, some are suggestions to (hopefully) improve the implementation ever so slightly. 

Overall this looks very nice, and it solves a huge problem of parsing nested content! 

**Edit:** 

> The scanner creates the vector of Symbols using the grapheme segmenter of icu.
This segmenter itself is an iterator, so we might be able to directly embed this iterator
in the SymbolIteratorRoot. This would require some form of a symbol cache to allow peeking,
but using an internal VecDeque should do the trick.

Regarding this, inspired by this I'm working on a crate called [`ribbon`](https://crates.io/crates/ribbon) that should help with this. It's an abstraction driven by an iterator, allowing for simultaneous access over multiple items yielded by an iterator at the same time. This could be useful at multiple places (iterating over symbols, inlines tokenization, inlines token resolver etc.).

File: commons/src/scanner/mod.rs
Comment (mhatzl) [reply commons/src/scanner/mod.rs:9]: I think `SymbolPosition` would be better in that case, and I would also change `Offset` to `SymbolOffset`.

File: commons/src/scanner/symbol/iterator/matcher.rs
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/matcher.rs:72]: I did not know that about `NOTE`.
But would you also write `NOTE` in doc-comments?
Usually I write `**Note:**` in doc-comments, but because normal comments have no formatting, I stayed with `Note` here, because it felt more consistent.

Would have to look through the code so replace all `Note`. Maybe keeping it as is for now, and change all in a new PR?

File: commons/src/scanner/symbol/iterator/root.rs
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/root.rs:78]: I replaced it with 

```
sym = ...filter(accept)?;
incr peek index;
some(sym)
```

File: parser/src/elements/atomic/heading.rs
Comment (mhatzl) [reply parser/src/elements/atomic/heading.rs:119]: yes I know. The *iterators all the way* approach is intended to solve this.

File: commons/src/scanner/symbol/iterator/mod.rs
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/mod.rs:11]: I would re-export `itertools`, because I assume mostly our `unimarkup` crates will depend on `commons`.
This makes it easy to keep the dependency of `itertools` in `commons`,
and all our other crates using `commons` automatically get features available in `itertools`.
Having to manage `itertools` in every crate using `commons` seems more painful.

File: commons/src/scanner/symbol/iterator/root.rs
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/root.rs:44]: Had to use `T: Into<&'input [Symbol<'input>]>`, but then I was able to combine both `From<>` implementations.

File: commons/src/scanner/mod.rs
Comment (mhatzl) [reply commons/src/scanner/mod.rs:25]: I decided to fully remove `Scanner`.
Instead I made `scan_str()` a public function in the `scanner` module.

File: commons/src/scanner/symbol/iterator/mod.rs
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/mod.rs]: added comparison for `start_index`, because this is the closest and most cheapest equality check we can get. Because we clone the iterator on `nest()`, other indices will most likely differ.

File: commons/src/scanner/symbol/iterator/matcher.rs
Comment (nfejzic) [reply commons/src/scanner/symbol/iterator/matcher.rs:72]: Hmm good question. Generally I wouldn't write any form of `Note:` in doc-comments. Would probably just explain it, something like `Note that ...`.

You can decide if and when you want to change this, it's not important part of this PR anyway.

Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/matcher.rs:72]: I think notes in doc-comments can be useful.
Think of them more like [GitHub alerts](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts).

It helps to highlight information that is especially relevant to a user.
Making a section bold does not make it readable, so you use alerts to highlight those sections instead.

File: commons/src/scanner/symbol/iterator/mod.rs
Comment (nfejzic) [reply commons/src/scanner/symbol/iterator/mod.rs:11]: There's also discussion about compatibility if we use different versions of `itertools` in for example `parser` and in `commons`. This solves that, but also makes it so that public api of `commons` is extended by that of `itertools`. I don't know what is best, so we can leave it as is for now.

File: commons/src/scanner/mod.rs
Comment (nfejzic) [reply commons/src/scanner/mod.rs:10]: `SymbolPosition` is looooong 😆. It does read better though. Both options are fine for me, you're free to choose whatever you find better 👍🏻.

P.S. if you can't choose, then choose randomly 🤣

File: commons/src/scanner/symbol/iterator/root.rs
Comment (nfejzic) [inline commons/src/scanner/symbol/iterator/root.rs:52]: Hahah, nice! I would say this will be a fun easter egg for future readers. I bet most of the people reading this will wonder what is this and learn something new 🤣

File: commons/src/scanner/symbol/iterator/root.rs
Comment (nfejzic) [inline commons/src/scanner/symbol/iterator/root.rs:63]: Much cleaner, nice!

File: commons/src/test_runner/mod.rs
Comment (nfejzic) [inline commons/src/test_runner/mod.rs:14]: Yup, exposing function for scanning was obvious when looking at this. Nice work 👏🏻

**Review by nfejzic — DISMISSED** [review]
Nice!

File: commons/src/scanner/mod.rs
Comment (mhatzl) [reply commons/src/scanner/mod.rs:10]: Or we keep the names and move them into the `symbol` module?
I was thinking about this option, but then `scanner` becomes a bit useless?
But removing scanner, by moving `symbol` up did not seem right to me.

File: commons/src/scanner/symbol/iterator/root.rs
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/root.rs:52]: 😄 it probably will be an easter egg for me in a year, but it looks nice.

**Comment by mhatzl** [comment]
I will wait for PR #105 to be merged, and then merge this one to resolve the clippy warnings that keep failing the action.

File: commons/src/scanner/symbol/iterator/matcher.rs
Comment (nfejzic) [reply commons/src/scanner/symbol/iterator/matcher.rs:72]: I know, but just in general keep in mind that it's not necessary most of the time. If something is that important, maybe separate heading is a better option. Otherwise we can just explain it. I also use `NOTE:` often, but we should probably reserve it for special cases. It kind of loses it's purpose if we over-use it, just want us to be aware of that.

**Comment by mhatzl** [comment]
I now merged changes from PR #105.
There were some merge conflicts, but all tests are passing now, and it should be ok to merge this PR.

For lexer tests in inline I had to remove the `EOI` symbol before running the tests.
With the `EOI` symbol, all lexer tests failed with an `unwrap` on `None`.
Assuming that we will update all of inline parsing to be based on iterators, I would leave this *workaround* as is for now.
Because inlines for now won't see the `EOI` symbol, because block elements strip it for them.

**Review by nfejzic — APPROVED** [review]

**Pull request merged**
